Allow passing of TrainSplit (train_test_split before) as parameter

dnouri opened the pull request:
An alternative to "train_test_splitter done using composition" #118

cancan101 commented:
Does this break current implementations that subclass and change `train_split`?

dnouri commented:
Good catch.  :-)

cancan101 commented on nolearn/lasagne/base.py:
```diff
@@ -83,6 +83,23 @@ def __getstate__(self):
         return state
 
 
+def train_split(X, y, net, eval_size=0.2):
```
Why not make this a class rather than a function so that eval_size can't be closed over rather than being passed in on each split.

dnouri replied:
Because the goal here is to make the interface easy to grasp to reuse, and a function is easier than a class.

dnouri commented:
I added b/w compatibility for NeuralNet subclasses with custom `train_test_split`.  Also added a bunch of tests.  At this point it's quite similar to #118, but adds a bunch of deprecation warnings and tests for b/w compatibility.

dnouri commented:
One positive side effect of using a function rather than the class interface that you propose, is that now you can treat `train_split_eval_size` or any other custom arguments to that function as an scikit-learn hyper parameter and do a parameter search using `net.set_params`.

Though thinking this through again, I'll have to take my "interface easier to grasp" argument back.  It's inconsistent with some other parts of the API.  And both class and function work at least equally well.  The `train_split_<param_name_here>` style of passing parameters might even trip some people up.  Now I tend to think that using a class is more natural after all.  :-)

dnouri commented:
Changed code to use a new TrainSplit class now, just like the class in #118.

This PR is in a slightly better shape though because:

- it adds tests for different backward compatibility cases
- it rips out the `train_test_split` method and thus makes a cleaner cut between old and new code

cancan101 commented on nolearn/lasagne/base.py:
```diff
@@ -117,6 +146,17 @@ def __init__(
+            train_split = LegacyTrainTestSplit(
+                eval_size=kwargs.pop('eval_size', 0.2))
+
+        if 'eval_size' in kwargs:  # BBB
```
Why would you be in this branch?

dnouri replied:
You passed `eval_size`, but have no custom subclass with `train_test_split`.  See this test: https://github.com/dnouri/nolearn/blob/feature-train-test-splitter/nolearn/lasagne/tests/test_base.py#L249-L251

cancan101 commented on nolearn/lasagne/base.py:
```diff
@@ -117,6 +146,17 @@ def __init__(
+
+        if 'eval_size' in kwargs:  # BBB
+            warn("The 'eval_size' argument has been deprecated, please use "
+                 "'train_split_eval_size' instead.")
```
What is `train_split_eval_size` ?

dnouri replied:
Good catch; that's from the commit before.

cancan101 commented on nolearn/lasagne/base.py:
```diff
@@ -83,6 +83,35 @@ def __getstate__(self):
         return state
 
 
+class TrainSplit:
```
Why not use new style class here.

dnouri replied:
Turned these to new-style (for Python 2) for consistency.
